### PR TITLE
Add truncated normal distribution

### PR DESCRIPTION
**Pull request by Erotemic**
I wanted a truncated normal parameter, but there wasn't one. So, I made one.

### CODE REVIEW

File: imgaug/parameters.py [hunk 1]
```diff
@@ -753,6 +753,74 @@ def __str__(self):
...
+        self.high = handle_continuous_param(high, "high")
+
+    def _draw_samples(self, size, random_state):
+        loc = self.loc.draw_sample(random_state=random_state)
+        scale = self.scale.draw_sample(random_state=random_state)
+        low = self.low.draw_sample(random_state=random_state)
+        high = self.high.draw_sample(random_state=random_state)
+        if low > high:
+            low, high = high, low
+        ia.do_assert(scale >= 0, "Expected scale to be in range [0, inf), got %s." % (scale,))
```
Comment (aleju) [inline imgaug/parameters.py]: Should add below
```python
if scale == 0:
    return np.full(size, loc, dtype=np.float64)
```
that's more efficient and without it you could get a division by zero.

File: imgaug/parameters.py [hunk 2]
```diff
@@ -753,6 +753,74 @@ def __str__(self):
...
+        a = (low - loc) / scale
+        b = (high - loc) / scale
+        rv = scipy.stats.truncnorm(a=a, b=b, loc=loc, scale=scale)
+        return rv.rvs(size=size, random_state=random_state)
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        return "Normal(loc=%s, scale=%s, low=%s, high=%s)" % (
```
Comment (aleju) [inline imgaug/parameters.py]: "TruncatedNormal(...)"

File: imgaug/parameters.py [hunk 3]
```diff
@@ -753,6 +753,74 @@ def __str__(self):
...
+    except the domain is smoothly bounded.
+
+    This is a wrapper around scipy.stats.truncnorm.
+
+    Parameters
+    ----------
+    loc : number or imgaug.parameters.StochasticParameter
+        The mean of the normal distribution.
+        If StochasticParameter, the mean will be sampled once per call
+        to :func:`imgaug.parameters.Normal._draw_samples`.
```
Comment (aleju) [inline imgaug/parameters.py]: typo: used here "Normal" instead of "TruncNormal"

File: imgaug/parameters.py [hunk 4]
```diff
@@ -753,6 +753,74 @@ def __str__(self):
...
+
+    scale : number or imgaug.parameters.StochasticParameter
+        The standard deviation of the normal distribution.
+        If StochasticParameter, the scale will be sampled once per call
+        to :func:`imgaug.parameters.Normal._draw_samples`.
```
Comment (aleju) [inline imgaug/parameters.py]: same as above

File: imgaug/parameters.py [hunk 5]
```diff
@@ -753,6 +753,74 @@ def __str__(self):
...
+
+    low : number or imgaug.parameters.StochasticParameter
+        The minimum value of the truncated normal distribution.
+        If StochasticParameter, the scale will be sampled once per call
+        to :func:`imgaug.parameters.Normal._draw_samples`.
```
Comment (aleju) [inline imgaug/parameters.py]: same as above

File: imgaug/parameters.py [hunk 6]
```diff
@@ -753,6 +753,74 @@ def __str__(self):
...
+
+    high : number or imgaug.parameters.StochasticParameter
+        The maximum value of the truncated normal distribution.
+        If StochasticParameter, the scale will be sampled once per call
+        to :func:`imgaug.parameters.Normal._draw_samples`.
```
Comment (aleju) [inline imgaug/parameters.py]: same as above

File: imgaug/parameters.py [hunk 7]
```diff
@@ -753,6 +753,74 @@ def __str__(self):
         return "Normal(loc=%s, scale=%s)" % (self.loc, self.scale)
 
 
+class TruncNormal(StochasticParameter):
```
Comment (aleju) [inline imgaug/parameters.py]: I think "TruncatedNormal" might be the better name. Not sure how many people are familiar with the scipy function, but many people are going to know the tensorflow one `truncated_normal` and are hence going to be more familiar with the non-abbreviated form. For anybody not knowing either of the two, the non-abbreviated form is also going to be more obvious.

**Comment by aleju** [comment]
Seems all good now. Merged this one. Thanks for the work!

Regarding ubelt, I think the __str__ and __repr__ functions are already implemented and tested, so replacing them with a different library would require a bit of work to get rid of that code. It might still need some custom str/repr functions and associated tests as not all augmenters always return all of their attributes in the produced strings, not all returned attributes necessarily inherit from StochasticParameter (e.g. lists of child augmenters) and sometimes the order of returned attributes is chosen to aid readability or reflect importance instead of matching the order in which they were initialized. Meanwhile, there will probably not be many new augmenters or parameters added in the next few months, so the workload associated with that will be low.
There is also the problem that the library already has quite some dependencies and I had hoped to reduce that number somehow.

**Comment by Erotemic** [comment]
This library has a great API, and I've been using it fairly extensively, but I agree it does have some issues. I understand about wanting to reduce bloat, but when I looked at your requirements file, it seems pretty tame. Is there anything specific that you were thinking about / is there any dependency that is unstable / difficult to install / the library would be better without?

The main issues / places for improvement I've seen in this library are: 

* The speed of augmenters can be slow, there is probably was to speed a lot of them up. 
* Data structures for bounding boxes, heatmaps, and keypoints are inefficient. (I can probably help with this. I have very efficient data structures in an internal code repo that is awaiting public release. The key idea behind them is to use numpy.ndarray / torch.tensor representations a collection of objects, which is much more efficient than holding a list of Python objects).
* It would be nice if multiple sequential AffineWarp operations could be fused in to a single operation for speed and numeric stability. I have absolutely no idea on how to do this, but I think `fastai` may have done something similar. 
* I think the way that parameters are registered could be improved to use a technique similar to the one used by pytorch to set attribute of its torch.Modules. Similar improvements could be made to the introspective abilities of the augmenters. This may also make it easier to serialize the augmenters as JSON, which would be really nice for logging hyperparameters of deep learning experiments (my netharn library has a hacky way to coerce imgaug objects to json, but its limited by the introspective capability of the API itself, so I'd really love to see an improvement there.)  

Let me know your thoughts on these issues / proposals. I'm more than happy to help with this work as well.

**Comment by aleju** [comment]
> I understand about wanting to reduce bloat, but when I looked at your requirements file, it seems pretty tame. Is there anything specific that you were thinking about / is there any dependency that is unstable / difficult to install / the library would be better without?

Shapely is currently only used for Polygons and these can't even be augmented yet. So it feels rather wasteful to have it as an additional dependency. `six` is I think only used for the range/xrange problem -- then again, `six` seems to have never caused problems for anybody.
Having many dependencies can cause problems. Anytime a base library changes from an update, that may lead to imgaug being broken. E.g. the last numpy update led to some modules of scikit-image being broken, which then led to travis failing here, which led to me having to spend an hour debugging the problem and dealing with it.
Some people might also not be able to install libraries at all or in specific configurations, e.g. due to their work environment, outdated/badly configured systems, Windows causing issues etc. These people might be lost users if they can't install the dependency. Or they might encounter problems during the installation of a dependency and think that it is caused by imgaug, leave an issue here and then I have to invest time into their problem. I recall such issues for opencv, scikit-image and shapely. Even if they are few in percent of the total number of users, they still create workload. That doesn't mean that adding dependencies is impossible, but there is some cost associated with it.

> The speed of augmenters can be slow, there is probably was to speed a lot of them up.

That is true. There is a branch ongoing improving the performance of some augmenters. It might get merged around end of january -- I hope.

> Data structures for bounding boxes, heatmaps, and keypoints are inefficient. 

Keypoints and Bounding Boxes are represented per image during augmentation as lists of Keypoint instances. That is indeed inefficient -- raw numpy arrays would be more efficient. That said, it still runs at several thousand keypoints per second in a standard scenario (batch size 128, 10 points per image), e.g. here with 60k per second in Affine and 880k per second in Fliplr/Flipud:
> KPs | #points=10 (on (224, 224)) B=128 FG | SUM    0.14539s | ITER avg    0.00145s, min    0.00133s, max    0.00178s | kps/s  880391.913 | mbit/s   537.349, mbyte/s    67.169 | Fliplr
> KPs | #points=10 (on (224, 224)) B=128 FG | SUM    0.14666s | ITER avg    0.00147s, min    0.00137s, max    0.00225s | kps/s  872787.694 | mbit/s   532.707, mbyte/s    66.588 | Flipud
> KPs | #points=10 (on (224, 224)) B=128 FG | SUM    2.13381s | ITER avg    0.02134s, min    0.02070s, max    0.02594s | kps/s   59986.515 | mbit/s    36.613, mbyte/s     4.577 | Affine_order_0_constant
> KPs | #points=10 (on (224, 224)) B=128 FG | SUM    2.12502s | ITER avg    0.02125s, min    0.02071s, max    0.02344s | kps/s   60234.726 | mbit/s    36.764, mbyte/s     4.596 | Affine_order_1_constant
> KPs | #points=10 (on (224, 224)) B=128 FG | SUM    2.13738s | ITER avg    0.02137s, min    0.02074s, max    0.03024s | kps/s   59886.293 | mbit/s    36.552, mbyte/s     4.569 | Affine_order_3_constant
> KPs | #points=10 (on (224, 224)) B=128 FG | SUM    2.10895s | ITER avg    0.02109s, min    0.02064s, max    0.02266s | kps/s   60693.795 | mbit/s    37.045, mbyte/s     4.631 | Affine_order_1_edge
> KPs | #points=10 (on (224, 224)) B=128 FG | SUM    2.12519s | ITER avg    0.02125s, min    0.02078s, max    0.02507s | kps/s   60230.003 | mbit/s    36.761, mbyte/s     4.595 | Affine_order_1_constant_skimage

You could of course convert in `augment_keypoints()` the input Keypoint instances to an array and then convert back in the end, which would still speed things up and it would be a change of rather manageable complexity.

Regarding the heatmaps, they seem currently comparably efficient? It is one numpy array for per image, with N channels for N sub-heatmaps. The segmentation maps on the other hand are quite inefficient as they are internally represented as heatmaps (one channel per class). I thought about changing them to purely integer-based arrays, but that is going to be a large change as all augmenters that affect somehow the geometry of input arrays would have to get a method to augment segmentations maps and each such method then would have to be unittested.

> It would be nice if multiple sequential AffineWarp operations could be fused in to a single operation for speed and numeric stability.

Currently the assumption is that people use the library for image augmentation and in that case a single affine transformation will almost always be enough. I can see additional ones being added after other augmentation techniques were applied (e.g. affine translation -> Crop -> affine rotation -> pad -> affine shear -- would be a very different effect than affine translation + rotation + shear -> crop -> pad). But in these cases you would not be able to combine the matrices.
In cases where you can combine them, you would have to change any augmenter that can contain other augmenters as children so that it automatically collapses multiple consecutive affine transformations (and handles each augmenters stochasticity and also does that without messing up the random states). That's a lot of additional complexity that is introduced and has to be tested for an ok-ish speedup in a rather uncommon situation.

> I think the way that parameters are registered could be improved to use a technique similar to the one used by pytorch to set attribute of its torch.Modules. Similar improvements could be made to the introspective abilities of the augmenters.

I'm not very familiar with the pytorch code, so I can't say anything about that. If it is purely about JSON, it would not be very hard to add `to_json()` methods to StochasticParameter subclasses and then add a `to_json()` to `Augmenter` that handles strings, numbers, tuples/lists/dicts, sub-augmenters and calls `to_json()` on StochasticParameters. Though it sounds like you want a bit more than that.

**Pull request merged**